server/services/discovery.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Discovery module discoveres hosts based on API names."""
+
+from dataclasses import dataclass
+import itertools
+from typing import Dict, List, Union
+import yaml
+
+import lib.config as libconfig
+
+cfg = libconfig.get_config()
+
+
+class InvalidEndpointException(Exception):
+  pass
+
+
+class InvalidIngressConfig(Exception):
+  pass
+
+
+# By default, all endpoints are expected on localhost port 8081.
+DEFAULT_INGRESS_RULES = dict({cfg.API_ROOT: ['/*']})
+
+
+@dataclass
+class Endpoint:
+  name: str  # Reference of the endpoint, use only within this server.
+  path: str  # Path to a specific API defined by mixer.
+
+
+class Endpoints:
+  """Container for endpoint."""
+
+  def __init__(self, endpoints: List[Endpoint]):
+    self.endpoints = endpoints
+    self.endpoint_name_to_path = {e.name: e.path for e in self.endpoints}
```

Comment:
Maybe we always use path as the name (hence removing the name field)?

Reply:
hm, I intentionally separated the label of an API from the path of an API. 

My understanding is that services.datacommons is supposed to be the service layer between raw DC api and  python interface of the website backends. I noticed that there are various places in the codebase that calls fetch_data, get, post, with urls outside of this module, breaking the service level abstraction. 

By
1) having all path encapsulated in the discovery module
2) having all data fetch / get / post encapsulated in services.datacommons module

We 
1) Make it clear what APIs in the mixer the website uses. You only need to look at discovery.py
2) Make our lives easier when mixer upgrades. For example, /v1/bulk/observations/point -> /v2/observations/point only only grades discovery.py. From the perspective of the rest of the code, the operation is still 'get_observations' so we can reference by label.
3) Can easily change the behaviour in services.datacommons.py. This layer handles API responses by calling the APIs.